ignite/metrics/nlp/bleu.py

```diff
@@ -185,20 +203,78 @@ def _corpus_bleu(
         gm = bp * math.exp(math.fsum(s))
         return gm
 
+    def _sentence_bleu(self, references: Sequence[Sequence[Any]], candidates: Sequence[Any],) -> float:
+        return self._corpus_bleu([references], [candidates])
+
+    def _corpus_bleu(
+        self, references: Sequence[Sequence[Sequence[Any]]], candidates: Sequence[Sequence[Any]],
+    ) -> float:
+
+        p_numerators: Counter = Counter()
+        p_denominators: Counter = Counter()
+
+        hyp_length_sum, ref_length_sum = self._n_gram_counter(
+            references=references, candidates=candidates, p_numerators=p_numerators, p_denominators=p_denominators,
+        )
+
+        bleu_score = self._brevity_penalty_smoothing(
+            p_numerators=p_numerators,
+            p_denominators=p_denominators,
+            hyp_length_sum=hyp_length_sum,
+            ref_length_sum=ref_length_sum,
+        )
+
+        return bleu_score
+
     @reinit__is_reduced
     def reset(self) -> None:
-        self._sum_of_bleu = torch.tensor(0.0, dtype=torch.double, device=self._device)
-        self._num_sentences = 0
+
+        if self.average == "macro":
+            self._sum_of_bleu = torch.tensor(0.0, dtype=torch.double, device=self._device)
+            self._num_sentences = 0
+
+        if self.average == "micro":
+            self.p_numerators: Counter = Counter()
+            self.p_denominators: Counter = Counter()
+            self.hyp_length_sum = 0
+            self.ref_length_sum = 0
 
     @reinit__is_reduced
     def update(self, output: Tuple[Sequence[Sequence[Any]], Sequence[Sequence[Sequence[Any]]]]) -> None:
         y_pred, y = output
-        for _y_pred, _y in zip(y_pred, y):
-            self._sum_of_bleu += self._corpus_bleu(references=[_y], candidates=[_y_pred])
-            self._num_sentences += 1
+
+        if self.average == "macro":
+            for refs, hyp in zip(y, y_pred):
+                self._sum_of_bleu += self._sentence_bleu(references=refs, candidates=hyp)
+                self._num_sentences += 1
+
+        elif self.average == "micro":
+            hyp_lengths, ref_lengths = self._n_gram_counter(
+                references=y, candidates=y_pred, p_numerators=self.p_numerators, p_denominators=self.p_denominators
+            )
+            self.hyp_length_sum += hyp_lengths
+            self.ref_length_sum += ref_lengths
 
     @sync_all_reduce("_sum_of_bleu", "_num_sentences")
-    def compute(self) -> torch.Tensor:
+    def _compute_macro(self) -> torch.Tensor:
         if self._num_sentences == 0:
             raise NotComputableError("Bleu must have at least one example before it can be computed.")
+
         return self._sum_of_bleu / self._num_sentences
+
+    @sync_all_reduce("p_numerators", "p_denominators", "hyp_length_sum", "ref_length_sum")
```

Comment:
I realized that a last thing should be done to work in ddp.

The original implementation is based on the Counter type. Here, it can’t work because it can’t be reduced by the decorator. We must replace Counter by torch.Tensor.

The p_numerators and p_denominators are counters of ngram. It could easily be replaced by a tensor of size ngram_order. The smoothers have to be updated.

Reply:
Yup I noticed this test failing, I was working on this only :)
Have added the changes.